velox/exec/tests/ArrowStreamTest.cpp

```diff
@@ -0,0 +1,240 @@
+ * limitations under the License.
+ */
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec::test;
+
+class ArrowStreamTest : public OperatorTestBase {
+ protected:
+  /// A mock Arrow reader returns an Arrow array for each call of getNext.
+  class ArrowReader {
+   public:
+    ArrowReader(
+        std::shared_ptr<memory::MemoryPool> pool,
+        const std::vector<RowVectorPtr>& vectors,
+        const TypePtr& type,
+        bool markGettingNextFailed = false /* For test only.*/,
+        bool markGettingSchemaFailed = false /* For test only.*/)
+        : pool_(pool),
+          vectors_(vectors),
+          type_(type),
+          markGettingNextFailed_(markGettingNextFailed),
+          markGettingSchemaFailed_(markGettingSchemaFailed) {}
+
+    int getNext(struct ArrowArray* outArray) {
+      if (vectorIndex_ < vectors_.size()) {
+        exportToArrow(vectors_[vectorIndex_], *outArray, pool_.get());
+        vectorIndex_ += 1;
+        outArray->release = bridgeRelease;
+      } else {
+        // End of stream. Mark the array released.
+        outArray->release = nullptr;
+      }
+      return markGettingNextFailed_ ? 1 : 0;
+    }
+
+    int getArrowSchema(ArrowSchema& out) {
+      exportToArrow(BaseVector::create(type_, 0, pool_.get()), out);
+      out.release = bridgeSchemaRelease;
+      return markGettingSchemaFailed_ ? 2 : 0;
+    }
+
+   private:
+    std::shared_ptr<memory::MemoryPool> pool_;
+    std::vector<RowVectorPtr> vectors_;
+    TypePtr type_;
+    int vectorIndex_ = 0;
+    bool markGettingNextFailed_;
+    bool markGettingSchemaFailed_;
+  };
+
+  /// For the export from Arrow reader to Arrow array stream.
+  class ExportedArrayStream {
+   public:
+    struct PrivateData {
+      explicit PrivateData(std::shared_ptr<ArrowReader> reader)
+          : reader_(std::move(reader)) {}
+
+      std::shared_ptr<ArrowReader> reader_;
+      std::string lastError_;
+
+      PrivateData() = default;
+    };
+
+    explicit ExportedArrayStream(struct ArrowArrayStream* stream)
+        : stream_(stream) {}
+
+    int getSchema(struct ArrowSchema* outSchema) {
+      return reader()->getArrowSchema(*outSchema);
+    }
+
+    int getNext(struct ArrowArray* outArray) {
+      return reader()->getNext(outArray);
+    }
+
+    const char* getLastError() {
+      const auto& lastError = privateData()->lastError_;
+      return lastError.empty() ? nullptr : lastError.c_str();
+    }
+
+    void release() {
+      if (stream_->release == nullptr) {
+        // Array stream is released.
+        return;
+      }
+      VELOX_CHECK_NOT_NULL(privateData());
+      delete privateData();
+      // Mark stream as released.
+      stream_->release = nullptr;
+    }
+
+    static int staticGetSchema(
```

Comment:
Should these methods be private?

Reply:
These static methods are called outside of class, so keep them public.